### PR TITLE
Performance and Sharding of the Entity Cluster Bot

### PR DESCRIPTION
**Pull request by nfmelendez**
- Improve performance in the entity clustering finding using the ego graph technique
- Implement Sharding with a shared graph that is updated using an aws dynamodb mutex and stored in aws s3
- Improve perfomance avoiding RPC calls and adding isContract cache. Add confiigurable timeouts
- Update networkx library
- Add Experimental entity graph viewer (graphql + d3.js)
-  Update forta agent to lastest both python and node-
- Add sharding tests

### CODE REVIEW

File: entity-cluster-bot/src/persistance.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import bz2
+import networkx as nx
+import time
+
+
+from boto3.dynamodb.conditions import Key, Attr
+
+
+try:
+    from src.keys import AWS_ACCESS_KEY
```
Comment (christian-forta) [inline entity-cluster-bot/src/persistance.py:16]: we would want to put all your secrets keys into a secrets json, which you can access using jwt. I can work with steven on that as we probably want to setup the infra under the forta foundation account.

File: entity-cluster-bot/README.md [hunk 2]
```diff
@@ -12,7 +12,104 @@ Further, contracts created by an address are also considered part of the same en
...
+The bot has a graph of the connections between the addresses that must be shared between many intances. All instances read and write the shared graph at the same time, so to avoid race condition it use a 
+mutex implemented in dynamodb using  [DynamoDb conditional write](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithItems.html#WorkingWithItems.ConditionalUpdate) and [atomic compare-and-swap](https://en.wikipedia.org/wiki/Compare-and-swap) 
+
+Also we store the graph in s3 as the graph compressed can be around 15MB and optimize costs
+
+Each instance updates the shared graoh every TX_SAVE_STEP 
+
+
+
+## Infrastructure
```
Comment (christian-forta) [inline entity-cluster-bot/README.md:28]: @stevenlanders , is this sufficient information for you to setup the appropriate AWS infra pieces?

File: entity-cluster-bot/src/agent.py [hunk 3]
```diff
@@ -1,335 +1,363 @@
...
+    from persistance import DynamoPersistance
+
 
 from os import environ
 
-from src.storage import get_secrets
 from dotenv import load_dotenv
 load_dotenv()
 
-SECRETS_JSON = get_secrets()
```
Comment (christian-forta) [inline entity-cluster-bot/src/agent.py:22]: this is how we can get secret keys. lets change it back. The format of the json is simple. Looks like this:

{
	"aws": {
		"key": "value"
         }
}

File: entity-cluster-bot/src/agent.py [hunk 4]
```diff
@@ -1,335 +1,363 @@
...
+
+    def create_finding(self, from_, message) -> Finding:
+        shared_graph = self.persistance.graph_cache
+        # Add all nodes and all edges to the shared cached graph to have a updated graph with the delta + shared
+        shared_graph.update(self.GRAPH)
+        filtered_graph = nx.subgraph_view(shared_graph, filter_edge=EntityClusterAgent.filter_edge(shared_graph))
+        checksum_addr = Web3.toChecksumAddress(from_)
+
+        ############################################
+        # Uncomment to view graph in cytoscape viewer
```
Comment (christian-forta) [inline entity-cluster-bot/src/agent.py:276]: this is pretty neat :)

File: entity-cluster-bot/src/agent_test.py [hunk 5]
```diff
@@ -1,108 +1,107 @@
...
-    
+        test_util_persistance = DynamoPersistance()
+        test_util_persistance.clean_db()
 
     def test_prune_graph_age(self):
         #  create a graph with some addresses which are older and newer than MAX_AGE_IN_DAYS
         #  assert that the old ones are removed and the new ones are not
         TestEntityClusterBot.remove_persistent_state()
-        agent.initialize()
+        entity_cluster_agent = EntityClusterAgent(DynamoPersistance())
```
Comment (christian-forta) [inline entity-cluster-bot/src/agent_test.py:28]: does the test write to the production DB?

**Review by christian-forta — CHANGES_REQUESTED** [review]

File: entity-cluster-bot/src/agent.py (hunk 3 above)
Comment (nfmelendez) [reply entity-cluster-bot/src/agent.py:22]: sure

File: entity-cluster-bot/src/agent_test.py (hunk 5 above)
Comment (nfmelendez) [reply entity-cluster-bot/src/agent_test.py:28]: yes, will improve: 

1. In s3 it writes to production using a different key to not overwrite production files
2. In the metadata dynamo it uses different tables that is ok
3. it does use the mutex dynamo db of production, that is wrong.

File: entity-cluster-bot/src/persistance.py (hunk 1 above)
Comment (nfmelendez) [reply entity-cluster-bot/src/persistance.py]: done

File: entity-cluster-bot/src/agent_test.py (hunk 5 above)
Comment (nfmelendez) [reply entity-cluster-bot/src/agent_test.py:27]: Now the test modify DEV database.  It still has both, dev and prod the same s3 but the test save with different bucket key so they don't overwrite

**Review by christian-forta — APPROVED** [review]
thank you for making those changes. Let me deploy it on Monday and see whether it plays nice with our infra.

**Pull request merged**